proxyd/consensus_tracker.go

```diff
@@ -21,101 +26,311 @@ type ConsensusTracker interface {
+	return ct.state.Safe
 }
 
 func (ct *InMemoryConsensusTracker) SetSafeBlockNumber(blockNumber hexutil.Uint64) {
 	defer ct.mutex.Unlock()
 	ct.mutex.Lock()
 
-	ct.safeBlockNumber = blockNumber
+	ct.state.Safe = blockNumber
 }
 
 func (ct *InMemoryConsensusTracker) GetFinalizedBlockNumber() hexutil.Uint64 {
 	defer ct.mutex.Unlock()
 	ct.mutex.Lock()
 
-	return ct.finalizedBlockNumber
+	return ct.state.Finalized
 }
 
 func (ct *InMemoryConsensusTracker) SetFinalizedBlockNumber(blockNumber hexutil.Uint64) {
 	defer ct.mutex.Unlock()
 	ct.mutex.Lock()
 
-	ct.finalizedBlockNumber = blockNumber
+	ct.state.Finalized = blockNumber
 }
 
-// RedisConsensusTracker uses a Redis `client` to store and retrieve consensus, async-safe
+// RedisConsensusTracker store and retrieve in a shared Redis cluster, with leader election
 type RedisConsensusTracker struct {
 	ctx          context.Context
 	client       *redis.Client
-	backendGroup string
+	namespace    string
+	backendGroup *BackendGroup
+
+	redlock           *redsync.Mutex
+	lockPeriod        time.Duration
+	heartbeatInterval time.Duration
+
+	leader     bool
+	leaderName string
+
+	// holds the state collected by local pollers
+	local *InMemoryConsensusTracker
+
+	// holds a copy of the remote shared state
+	// when leader, updates the remote with the local state
+	remote *InMemoryConsensusTracker
+}
+
+type RedisConsensusTrackerOpt func(cp *RedisConsensusTracker)
+
+func WithLockPeriod(lockPeriod time.Duration) RedisConsensusTrackerOpt {
+	return func(ct *RedisConsensusTracker) {
+		ct.lockPeriod = lockPeriod
+	}
 }
 
-func NewRedisConsensusTracker(ctx context.Context, r *redis.Client, namespace string) ConsensusTracker {
-	return &RedisConsensusTracker{
+func WithHeartbeatInterval(heartbeatInterval time.Duration) RedisConsensusTrackerOpt {
+	return func(ct *RedisConsensusTracker) {
+		ct.heartbeatInterval = heartbeatInterval
+	}
+}
+func NewRedisConsensusTracker(ctx context.Context,
+	redisClient *redis.Client,
+	bg *BackendGroup,
+	namespace string,
+	opts ...RedisConsensusTrackerOpt) ConsensusTracker {
+
+	tracker := &RedisConsensusTracker{
 		ctx:          ctx,
-		client:       r,
-		backendGroup: namespace,
+		client:       redisClient,
+		backendGroup: bg,
+		namespace:    namespace,
+
+		lockPeriod:        30 * time.Second,
+		heartbeatInterval: 2 * time.Second,
+		local:             NewInMemoryConsensusTracker().(*InMemoryConsensusTracker),
+		remote:            NewInMemoryConsensusTracker().(*InMemoryConsensusTracker),
+	}
+
+	for _, opt := range opts {
+		opt(tracker)
+	}
+
+	return tracker
+}
+
+func (ct *RedisConsensusTracker) Init() {
+	go func() {
+		for {
+			// follow same context as backend group poller
+			ctx := ct.backendGroup.Consensus.ctx
+			timer := time.NewTimer(ct.heartbeatInterval)
+			ct.stateHeartbeat()
+
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				timer.Stop()
+				return
+			}
+		}
+	}()
+}
+
+func (ct *RedisConsensusTracker) stateHeartbeat() {
```

Comment:
`stateHeartbeat` does a lot of logging but returns void - may be better to have `stateHeartbeat` return an error, and do the logging in the caller.

Reply:
I rather not.  This is a complex logic and it is more readable to have the logging in-place.  They help to understand the story behind the logic in a way the code is communicating to the outside world.